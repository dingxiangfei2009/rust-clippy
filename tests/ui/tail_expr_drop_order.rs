//@compile-flags: -Z unstable-options
//@edition:2024

#![warn(clippy::tail_expr_drop_order)]

struct LoudDropper;
impl Drop for LoudDropper {
    fn drop(&mut self) {
        println!("loud drop")
    }
}
impl LoudDropper {
    fn get(&self) -> i32 {
        0
    }
}

fn should_lint() -> i32 {
    let x = LoudDropper;
    // Should lint
    x.get() + LoudDropper.get() //~ ERROR: discretion required on this expression which generates a value with a significant drop implementation
}

fn should_not_lint() -> i32 {
    let x = LoudDropper;
    // Should not lint
    x.get()
}

fn main() {}
