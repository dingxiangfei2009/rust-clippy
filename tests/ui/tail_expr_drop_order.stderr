error: discretion required on this expression which generates a value with a significant drop implementation
  --> tests/ui/tail_expr_drop_order.rs:21:15
   |
LL |     x.get() + LoudDropper.get()
   |               ^^^^^^^^^^^
   |
help: one or more locals with a significant drop implementation will observe a visible change in drop order
  --> tests/ui/tail_expr_drop_order.rs:19:9
   |
LL |     let x = LoudDropper;
   |         ^
   = note: `-D clippy::tail-expr-drop-order` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::tail_expr_drop_order)]`

error: aborting due to 1 previous error

